src/tests/Microsoft.PowerFx.Interpreter.Tests/IRTests/RewritingIRVisitorTests.cs

```diff
@@ -239,5 +242,438 @@ public override IntermediateNode Visit(DecimalLiteralNode node, Context context)
+            "(LookUp(MyTable, (EqText(Name,Bread)))).Value",
+            "3.5")]
+        [InlineData(
+            "Last(MyTable).Name",
+            "Substitute(Substitute(Substitute((Last(MyTable)).Name, &, &amp;), <, &lt;), >, &gt;)",
+            "Mac &amp; cheese")]
+        [InlineData(
+            "With({a:'<p>' & StringValue & '</p>'}, a & a)",
+            "With({a:Concatenate(<p>, Substitute(Substitute(Substitute(StringValue, &, &amp;), <, &lt;), >, &gt;), </p>)}, (Concatenate(a, a)))",
+            "<p>&lt;script&gt;alert('here');&lt;/script&gt;</p><p>&lt;script&gt;alert('here');&lt;/script&gt;</p>")]
+        public void TestEscapingHtml(string expr, string expectedIR, string expectedResult)
+        {
+            var expression = expr.Replace("\'", "\"");
+            var engine = new RecalcEngine();
+            engine.UpdateVariable("StringValue", "<script>alert('here');</script>");
+            var tableRecordType = new KnownRecordType(TestUtils.DT("![Name:s,Value:n]"));
+            var tableVar = FormulaValue.NewTable(
+                tableRecordType,
+                new InMemoryRecordValue(
+                    IRContext.NotInSource(tableRecordType),
+                    new NamedValue("Name", FormulaValue.New("Bread")),
+                    new NamedValue("Value", FormulaValue.New(3.50))),
+                new InMemoryRecordValue(
+                    IRContext.NotInSource(tableRecordType),
+                    new NamedValue("Name", FormulaValue.New("Milk")),
+                    new NamedValue("Value", FormulaValue.New(4.99))),
+                new InMemoryRecordValue(
+                    IRContext.NotInSource(tableRecordType),
+                    new NamedValue("Name", FormulaValue.New("Cheese")),
+                    new NamedValue("Value", FormulaValue.New(5.75))),
+                new InMemoryRecordValue(
+                    IRContext.NotInSource(tableRecordType),
+                    new NamedValue("Name", FormulaValue.New("Mac & cheese")),
+                    new NamedValue("Value", FormulaValue.New(2.25))));
+            engine.UpdateVariable("MyTable", tableVar);
+
+            engine.IRTransformList.Add(new EscapeHtmlTransform());
+
+            var opts = new ParserOptions { AllowsSideEffects = true };
+            var check = new CheckResult(engine)
+                .SetText(expression, opts)
+                .SetBindingInfo();
+
+            var ir = check.ApplyIR(); // calls rewriter
+
+            var actualIR = check.GetCompactIRString();
+
+            Assert.Equal(expectedIR.Replace("\'", "\""), actualIR);
+
+            var run = check.GetEvaluator();
+            var result = run.Eval();
+
+            var objActual = result.ToObject();
+            if (objActual is not string)
+            {
+                objActual = objActual.ToString();
+            }
+
+            Assert.Equal(objActual, expectedResult);
+        }
+
+        private class EscapeHtmlTransform : IRTransform
```

Comment:
Great test. 
Looks like it's using the same infra we use for IRrewriting for delegation .

Reply:
👍